[MLIR][TF][SCF] Introduce tf.IfRegion -> scf.if lowering

srishti-pm opened the pull request:
Lowering for `tf.IfRegion` -> `scf.if` implemented in a new pass
`convert-tf-control-flow-to-scf`, which is added in
`tensorflow/transforms` and can be used for all direct control flow
lowerings from the TensorFlow dialect to the SCF dialect (without going
through dialects like HLO, which involve tuples, for example). Currently
this pass has one pattern `ConvertIfRegionOp`. It lowers the
`tf.IfRegion` op to `scf.if`.

Signed-off-by: Srishti Srivastava <srishti.srivastava@polymagelabs.com>

joker-eph commented:
FYI: using MHLO does not have to involve tuples (actually we'd like to remove them from MHLO entirely).

I'm curious about the plan for this new directory: what's the motivation and what about the redundancy with the other paths (through MHLO and TOSA)?

srishti-pm commented:
> Thanks for the contribution!
> 
> FYI: using MHLO does not have to involve tuples (actually we'd like to remove them from MHLO entirely).
> 
> I'm curious about the plan for this new directory: what's the motivation and what about the redundancy with the other paths (through MHLO and TOSA)?
> 
> Maybe this was discussed elsewhere and I missed it? (Feel free to post a pointer then)

Hi. A path through MHLO (which is currently using tuples) exists but it involves tupling and then detupling. So, the motivation for this PR is to have a direct path from the `tensorflow` dialect to the `SCF` dialect in order to avoid this overhead.

srishti-pm commented:
> > Thanks for the contribution!
> > FYI: using MHLO does not have to involve tuples (actually we'd like to remove them from MHLO entirely).
> > I'm curious about the plan for this new directory: what's the motivation and what about the redundancy with the other paths (through MHLO and TOSA)?
> > Maybe this was discussed elsewhere and I missed it? (Feel free to post a pointer then)
> 
> Hi. A path through MHLO (which is currently using tuples) exists but it involves tupling and then detupling. So, the motivation for this PR is to have a direct path from the `tensorflow` dialect to the `SCF` dialect in order to avoid this overhead.

@bondhugula , maybe you can elaborate more on this?

joker-eph commented:
Are you concerned with the runtime performance with the “overhead” here?
I’m a bit worried about redundant code paths in the codebase, which is why I’m trying to identify the elements to balance here.

srishti-pm commented:
> with the runtime performance with the “overhead” here?

Yes, runtime performance.

joker-eph commented:
I’m interested to look deeper into the MHLO path instead: do you have an example we can walk through?

srishti-pm commented:
> I’m interested to look deeper into the MHLO path instead: do you have an example we can walk through?

Sure. Let's look at the `tf.If` op in the function @testValidConvertIfOp1 of this PR.

bondhugula commented:
> Are you concerned with the runtime performance with the “overhead” here?
> I’m a bit worried about redundant code paths in the codebase, which is why I’m trying to identify the elements to balance here.

A `tf.if` to `scf.if` lowering (and similarly `tf.while` to `scf.while`) would appear to be the natural path towards code generation or further lowering without any abrupt drop in abstraction. Both the TF and SCF dialect ops on `if` and `while` do not use any tuples and there isn't a need to go through intermediate dialects. The lowering path going through TOSA could naturally make use of this as both TOSA and SCF dialects are present upstream - so you get a self-contained upstream target to convert these to. @silvasean had earlier reported on a similar conversion of Torch `if` ops to scf.if for example which was also natural.

bondhugula commented:
@srishti-pm Can you get rid of the merge commit from your PR and have your changes in a single commit?

srishti-pm commented:
@joker-eph, can you please review this PR?

joker-eph commented:
> @joker-eph, can you please review this PR?

Before looking at the code, I'm trying to wrap my head around the direction here. This PR does not fit into the existing paths (it is quite visible by adding a top-level directory in compiler/mlir), so I'm cautious here. I'll try to get back to this this week, but ping me on Monday if you don't hear back.

srishti-pm committed:
[MLIR][TF][SCF] Introduce tf.IfRegion -> scf.if lowering

Lowering for `tf.IfRegion` -> `scf.if` implemented in a new pass
`convert-tf-control-flow-to-scf`, which is added in
`tensorflow/transforms` and can be used for all direct control flow
lowerings from the TensorFlow dialect to the SCF dialect (without going
through dialects like HLO, which involve tuples, for example). Currently
this pass has one pattern `ConvertIfRegionOp`. It lowers the
`tf.IfRegion` op to `scf.if`.

Signed-off-by: Srishti Srivastava <srishti.srivastava@polymagelabs.com>

srishti-pm commented:
> > @joker-eph, can you please review this PR?
> 
> Before looking at the code, I'm trying to wrap my head around the direction here. This PR does not fit into the existing paths (it is quite visible by adding a top-level directory in compiler/mlir), so I'm cautious here. I'll try to get back to this this week, but ping me on Monday if you don't hear back.

Hi @joker-eph, pinging you for reviewing this!

joker-eph commented:
Ok, let's integrate this into `tensorflow/compiler/mlir/tensorflow/transforms/...` instead of a top-level here.
Also I think the code will be simple if you place it after the `-tf-functional-control-flow-to-regions`: the control flow will already be expressed with region.

srishti-pm commented:
> Ok, let's integrate this into `tensorflow/compiler/mlir/tensorflow/transforms/...` instead of a top-level here.
> Also I think the code will be simple if you place it after the `-tf-functional-control-flow-to-regions`: the control flow will already be expressed with region.

Sure, I'll do this.

srishti-pm commented:
@joker-eph , I have addressed your comments and modified my task goal and implementation accordingly.

1. Instead of converting `tf.If` to `scf.if`, I am converting `tf.IfRegion` to `scf.if`.
2. Instead of having a new directory for the `SCF` dialect, I have added my pass in `tensorflow/transforms`

joker-eph commented on tensorflow/compiler/mlir/tensorflow/tests/convert-tf-control-flow-to-scf.mlir:
```diff
@@ -0,0 +1,73 @@
+  // CHECK-NEXT: return
+}
+
+// `tf.IfRegion` doesn't get converted to `scf.if` because its return type(s) don't match the type(s) of the operand(s) of the `then` and `else` terminators.
```
Have you considered introducing a cast before the yield during the conversion to make the conversion work in all cases?

srishti-pm replied:
I'll do that. Thanks for the suggestion!

srishti-pm replied:
@joker-eph This is done now. The conversion happens in all cases.

srishti-pm commented:
@joker-eph , I have addressed all your comments given on this PR.

srishti-pm commented on tensorflow/compiler/mlir/tensorflow/transforms/convert_tf_control_flow_to_scf.cc:
```diff
@@ -0,0 +1,117 @@
+  return std::make_unique<ConvertTfControlFlowToScf>();
+}
+
+static PassRegistration<ConvertTfControlFlowToScf> pass;
```
Sure. I'll do this.

srishti-pm commented on tensorflow/compiler/mlir/tensorflow/transforms/convert_tf_control_flow_to_scf.cc:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<scf::SCFDialect, tensor::TensorDialect>();
```
Because I am using the `tensor.extract` op to convert the `tensor<i1>` condition of `tf.IfRegion` to the `i1` condition of `scf.if`. An `scf.if` op's condition is of type `i1` while that of a `tf.IfRegion` op is `tensor<i1>`.

srishti-pm replied:
Ohh... No problem ! :)

srishti-pm commented:
@joker-eph , the PR is ready for review again !

joker-eph commented:
CI is complaining right now with:
```
third_party/tensorflow/compiler/mlir/tensorflow/transforms/convert_tf_control_flow_to_scf.cc:49:12: error: unused variable 'scfYield' [-Werror,-Wunused-variable]
      auto scfYield = rewriter.create<scf::YieldOp>(currentTerminator->getLoc(),
           ^
third_party/tensorflow/compiler/mlir/tensorflow/transforms/convert_tf_control_flow_to_scf.cc:95:5: error: ignoring return value of function declared with 'nodiscard' attribute [-Werror,-Wunused-result]
    applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

srishti-pm commented:
@joker-eph @jpienaar, I have addressed all the comments given on the PR:
1. Modified the variable naming style.
2. Removed extraneous information from the pass description in the `.td` file.
